common/formatting/layout_optimizer_test.cc

```diff
@@ -0,0 +1,1265 @@
+      EXPECT_EQ(layout_tree.Children().size(), 2);
+      const auto& layout = layout_tree.Value();
+      EXPECT_EQ(layout.GetType(), LayoutType::kLayoutVerticalMerge);
+
+      const auto& layout_subtree_upper = layout_tree.Children()[0];
+      EXPECT_EQ(layout_subtree_upper.Children().size(), 0);
+      const auto& layout_upper = layout_subtree_upper.Value();
+      EXPECT_EQ(layout_upper.GetType(), LayoutType::kLayoutLine);
+
+      const auto& layout_subtree_lower = layout_tree.Children()[1];
+      EXPECT_EQ(layout_subtree_lower.Children().size(), 0);
+      const auto& layout_lower = layout_subtree_lower.Value();
+      EXPECT_EQ(layout_lower.GetType(), LayoutType::kLayoutLine);
+    }
+    EXPECT_EQ(knot.GetSpacesBefore(), 1);
+    EXPECT_EQ(knot.GetSpacingOptions(), SpacingOptions::Undecided);
+  }
+  {
+    const auto& knot = knot_set_wrapped[2];
+    EXPECT_EQ(knot.GetColumn(), 11);
+    EXPECT_EQ(knot.GetSpan(), 7);
+    EXPECT_EQ(knot.GetIntercept(), 510.002f);
+    EXPECT_EQ(knot.GetGradient(), 96);
+    {
+      const auto layout_tree = knot.GetLayout();
+      EXPECT_EQ(layout_tree.Children().size(), 2);
+      const auto& layout = layout_tree.Value();
+      EXPECT_EQ(layout.GetType(), LayoutType::kLayoutVerticalMerge);
+
+      const auto& layout_subtree_upper = layout_tree.Children()[0];
+      EXPECT_EQ(layout_subtree_upper.Children().size(), 0);
+      const auto& layout_upper = layout_subtree_upper.Value();
+      EXPECT_EQ(layout_upper.GetType(), LayoutType::kLayoutLine);
+
+      const auto& layout_subtree_lower = layout_tree.Children()[1];
+      EXPECT_EQ(layout_subtree_lower.Children().size(), 0);
+      const auto& layout_lower = layout_subtree_lower.Value();
+      EXPECT_EQ(layout_lower.GetType(), LayoutType::kLayoutLine);
+    }
+    EXPECT_EQ(knot.GetSpacesBefore(), 1);
+    EXPECT_EQ(knot.GetSpacingOptions(), SpacingOptions::Undecided);
+  }
+}
+
+class TreeReconstructorTestFixture : public ::testing::Test,
+                                     public UnwrappedLineMemoryHandler {
+ public:
+  TreeReconstructorTestFixture()
+      : sample_("first_line second_line third_line fourth_line"),
+        tokens_(absl::StrSplit(sample_, ' ')) {
+    for (const auto token : tokens_) {
+      ftokens_.emplace_back(TokenInfo{1, token});
+    }
+    CreateTokenInfos(ftokens_);
+  }
+
+ protected:
+  const std::string sample_;
+  const std::vector<absl::string_view> tokens_;
+  std::vector<TokenInfo> ftokens_;
+};
+
+TEST_F(TreeReconstructorTestFixture, SingleLine) {
+  const auto& preformat_tokens = pre_format_tokens_;
+  const auto begin = preformat_tokens.begin();
+  BasicFormatStyle style;
+
+  UnwrappedLine single_line(0, begin);
+  single_line.SpanUpToToken(begin + 1);
+
+  const auto layout_tree = LayoutTree{Layout{single_line}};
+  TreeReconstructor tree_reconstructor(0, style);
+  tree_reconstructor.TraverseTree(layout_tree);
+
+  auto optimized_tree = VectorTree{UnwrappedLine(0, begin)};
```

Comment:
This would need some deduction guide, with explicitly passing the type e.g.
`VectorTree<UnwrappedLine>{UnwrappedLine(0, begin)};`
... or maybe finding a good name and `using` it. 

(we're compiling with clang++ `-Wctad-maybe-unsupported`, which then warns about the vector tree maybe not intending class template deduction)

↓ also for all the other auto optimized_tree

Reply:
Done